### ISSUE TITLE
Low cache hit count for conda builds on Windows

### ISSUE THREAD

**Issue by peterjc123** [pytorch/pytorch#36381]
## Fact

Even when building for a conda release for a second time, the cache hit count is pretty low. Example result:
```
Compile requests               1978
Compile requests executed      1971
Cache hits                      221
Cache misses                   1739
Cache timeouts                    0
Cache read errors                 0
Forced recaches                   0
Cache write errors                0
Compilation failures              6
Cache errors                      5
Non-cacheable compilations       33
Non-cacheable calls               0
Non-compilation calls             7
Unsupported compiler calls        0
Average cache write           0.224 s
Average cache read miss      25.052 s
Average cache read hit        0.060 s
Cache location             S3, bucket: Bucket(name=ossci-compiler-cache-circleci-v2, base_url=http://ossci-compiler-cache-circleci-v2.s3.amazonaws.com/)
```

## Underlying bug
When calculating hashes, the switch `-E` is used, which contains the full path and the line number of the included headers, which will reduce the cache rate when we happen to make builds in different directories.

## Solution
Use `-EP` instead. Apply https://github.com/mozilla/sccache/pull/723. As for NVCC, `-Xcompiler -EP` should be passed.

cc @yf225

**Comment by peterjc123** [comment]
I attempted to fix that with https://github.com/peterjc123/sccache/tree/msvc_improve_cache_hits and the binary is generated at https://drive.google.com/u/0/uc?id=1WvPeBpn41hdr6gQ29uXiQ5RjlzeFvJbO&export=download.

**Comment by peterjc123** [comment]
@yf225 Would you please create a new S3 bucket for me to test the changes?

**Comment by yf225** [comment]
@peterjc123 Thanks a lot for digging into it! @ezyang Curious what's the current recommended way to create new S3 bucket for testing?

**Comment by peterjc123** [comment]
Sorry, the build is failing with some other reasons. So looks like we don't need to create a new S3 bucket.

**Comment by peterjc123** [comment]
Well, looks like that is still not enough. Maybe we should transform all the paths to relative path (can be specified using a environment variable) while calculating hashes.

**Comment by ezyang** [comment]
My personal preference is to just force the build to always happen at the same location, so then you don't have to transform anything. Is there a reason we can't conveniently do that here?

**Comment by peterjc123** [comment]
@ezyang Conda-build doesn't provide an option to pin the build directory. See https://github.com/conda/conda-build/blob/97d35e337419c08945acf29605791dc9e8d4eb81/conda_build/config.py#L551 for more details.

**Comment by ezyang** [comment]
RIP

**Comment by peterjc123** [comment]
Let's open an issue on conda-build first.

**Comment by peterjc123** [comment]
@ezyang I managed to pin the build directory with a few tricks. BTW, what you think of a new bucket for Windows nightly builds with CircleCI exclusively?

**Comment by ezyang** [comment]
> BTW, what you think of a new bucket for Windows nightly builds with CircleCI exclusively? Currently we are sharing the same one as the Linux binary builds.

Sure. sccache shouldn't clobber in any case, but if you want to bump up the cache size this would be a reasonable call.

**Comment by peterjc123** [comment]
With the help of a few tricks, I managed to fix this issue. The following is the report of the sccache with the fix.
```
Compile requests              1967
Compile requests executed     1960
Cache hits                    1915
Cache misses                    34
Cache timeouts                   0
Cache read errors                0
Forced recaches                  0
Cache write errors               0
Compilation failures             6
Cache errors                     5
Non-cacheable compilations      33
Non-cacheable calls              0
Non-compilation calls            7
Unsupported compiler calls       0
Average cache write          0.225 s
Average cache read miss      0.565 s
Average cache read hit       0.482 s
Cache location             S3, bucket: Bucket(name=ossci-compiler-cache-circleci-v2, base_url=http://ossci-compiler-cache-circleci-v2.s3.amazonaws.com/)
```

**Comment by peterjc123** [comment]
> if you want to bump up the cache size this would be a reasonable call.

Would you please do this?

**Comment by ezyang** [comment]
OK new bucket is called ossci-compiler-cache-windows

**Comment by peterjc123** [comment]
@ezyang Which AWS key should be used? I tried `CIRCLECI_AWS_ACCESS_KEY_FOR_SCCACHE_S3_BUCKET_V4` and `CIRCLECI_AWS_SECRET_KEY_FOR_SCCACHE_S3_BUCKET_V4` but the cache hits are zero after rebuild.

**Comment by peterjc123** [comment]
@ezyang Which AWS key should be used? I tried `CIRCLECI_AWS_ACCESS_KEY_FOR_SCCACHE_S3_BUCKET_V4` and `CIRCLECI_AWS_SECRET_KEY_FOR_SCCACHE_S3_BUCKET_V4` but the cache hits are zero after rebuild.

**Comment by ezyang** [comment]
Yes it looks like I need to issue a new access key

**Comment by ezyang** [comment]
OK, I think I added access to the new bucket from `CIRCLECI_AWS_ACCESS_KEY_FOR_SCCACHE_S3_BUCKET_V4`

**Comment by peterjc123** [comment]
> OK, I think I added access to the new bucket from `CIRCLECI_AWS_ACCESS_KEY_FOR_SCCACHE_S3_BUCKET_V4`

Let me try again.

**Comment by peterjc123** [comment]
@ezyang They seem to be working.

**Comment by peterjc123** [comment]
> OK, I think I added access to the new bucket from `CIRCLECI_AWS_ACCESS_KEY_FOR_SCCACHE_S3_BUCKET_V4`

I don't know why but the cache hit rate is still 0%. The failed cache write count is nearly zero. So maybe you forget to enable the read access for the token?

**Comment by ezyang** [comment]
Can you post a log? It looks like read access is allowed (but I might be looking at the IAM wrong). If reads aren't working, then there should be some stats from sccache saying that we can't find anything.

**Comment by peterjc123** [comment]
@yf225 Do you have idea on how to produce a log for sccache with S3?

**Comment by peterjc123** [comment]
@ezyang Here is the debug log provided by sccache.
```
DEBUG:sccache::cache::cache: Trying S3Cache(ossci-compiler-cache-windows.s3.amazonaws.com)
DEBUG:sccache::server: handle_client: zero_stats
DEBUG:sccache::server: handle_client: compile
DEBUG:sccache::compiler::compiler: executable: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe
DEBUG:sccache::compiler::compiler: executable: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe
DEBUG:sccache::compiler::compiler: Found MSVC
DEBUG:sccache::server: check_compiler: Supported compiler
DEBUG:sccache::server: parse_arguments: Ok: ["/EHSc", "/c", "a.cpp"]
DEBUG:sccache::compiler::compiler: [a.obj]: get_cached_or_compile: ["/EHSc", "/c", "a.cpp"]
DEBUG:sccache::compiler::msvc: preprocess: Some("C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\cl.exe" "-E" "a.cpp" "-nologo" "-EHSc")
DEBUG:sccache::compiler::compiler: [a.obj]: generate_hash_key took 0.071 s
DEBUG:sccache::simples3::s3: GET http://ossci-compiler-cache-windows.s3.amazonaws.com/7/b/6/7b62ce3178856366e6dbe775ce59b3fd62d44e1b8536bed35b08472e6840725c4d4b6d7e8f9d8016ebd3d84b9053a8fffa9da64eee6c1ab410597dac9c813699
WARN:sccache::cache::s3: Got AWS error: Error(BadHTTPStatus(Forbidden), State { next_error: None })
DEBUG:sccache::compiler::compiler: [a.obj]: Cache miss in 1.760 s
DEBUG:sccache::compiler::compiler: [a.obj]: Compiled in 0.101 s, storing in cache
DEBUG:sccache::simples3::credential: Using AWS credentials from environment
DEBUG:sccache::simples3::s3: PUT http://ossci-compiler-cache-windows.s3.amazonaws.com/7/b/6/7b62ce3178856366e6dbe775ce59b3fd62d44e1b8536bed35b08472e6840725c4d4b6d7e8f9d8016ebd3d84b9053a8fffa9da64eee6c1ab410597dac9c813699
DEBUG:sccache::compiler::compiler: [a.obj]: Stored in cache successfully!
DEBUG:sccache::server: [a.obj]: Cache write finished in 0.494 s
DEBUG:sccache::server: handle_client: compile
DEBUG:sccache::server: check_compiler: Supported compiler
DEBUG:sccache::server: parse_arguments: Ok: ["/EHSc", "/c", "a.cpp"]
DEBUG:sccache::compiler::compiler: [a.obj]: get_cached_or_compile: ["/EHSc", "/c", "a.cpp"]
DEBUG:sccache::compiler::msvc: preprocess: Some("C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\cl.exe" "-E" "a.cpp" "-nologo" "-EHSc")
DEBUG:sccache::compiler::compiler: [a.obj]: generate_hash_key took 0.223 s
DEBUG:sccache::simples3::s3: GET http://ossci-compiler-cache-windows.s3.amazonaws.com/7/b/6/7b62ce3178856366e6dbe775ce59b3fd62d44e1b8536bed35b08472e6840725c4d4b6d7e8f9d8016ebd3d84b9053a8fffa9da64eee6c1ab410597dac9c813699
WARN:sccache::cache::s3: Got AWS error: Error(BadHTTPStatus(Forbidden), State { next_error: None })
DEBUG:sccache::compiler::compiler: [a.obj]: Cache miss in 1.660 s
DEBUG:sccache::compiler::compiler: [a.obj]: Compiled in 0.097 s, storing in cache
DEBUG:sccache::simples3::s3: PUT http://ossci-compiler-cache-windows.s3.amazonaws.com/7/b/6/7b62ce3178856366e6dbe775ce59b3fd62d44e1b8536bed35b08472e6840725c4d4b6d7e8f9d8016ebd3d84b9053a8fffa9da64eee6c1ab410597dac9c813699
DEBUG:sccache::compiler::compiler: [a.obj]: Stored in cache successfully!
DEBUG:sccache::server: [a.obj]: Cache write finished in 0.497 s
```

**Comment by peterjc123** [comment]
@ezyang It seems that the cache is not read using AWS credentials so you must set the acl of the bucket to public-read.

**Comment by peterjc123** [comment]
sccache from the upstream doesn't have this problem, so I tried to update the sccache with https://github.com/yf225/sccache/pull/2. However, there are currently some bugs including
1. NVCC undetected on Windows (easy to fix)
2. Some arguments starting with `/` is uncacheable.

**Comment by peterjc123** [comment]
New binary fixed this issue and is verified in https://github.com/pytorch/pytorch/pull/36719.

**Comment by peterjc123** [comment]
However, it seems that the hash calculation logic is changed so it leads to high cache miss time. 
```
Average cache read miss           36.759 s
```
@ezyang Would you please purge the cache in the bucket `ossci-compiler-cache-windows` and `ossci-compiler-cache`?

**Comment by peterjc123** [comment]
Sorry for the noise. There's no need to purge the cache.

**Comment by ezyang** [comment]
@peterjc123 You're right, thank you for debugging this. I've made the bucket have public read access too now. I'll go ahead and purge the cache too.
